File: zig/nucleotide-count/nucleotide_count.zig
pub const NucleotideError = error{Invalid};

pub const Counts = struct {
    a: u32,
    c: u32,
    g: u32,
    t: u32,
};

pub fn countNucleotides(s: []const u8) NucleotideError!Counts {
    var counts = Counts{ .a = 0, .c = 0, .g = 0, .t = 0 };

    for (s) |nucleotide| {
        switch (nucleotide) {
            'A' => counts.a += 1,
            'C' => counts.c += 1,
            'G' => counts.g += 1,
            'T' => counts.t += 1,
            else => return NucleotideError.Invalid,
        }
    }

    return counts;
}

File: julia/nucleotide-count/nucleotide-count.jl
function count_nucleotides(strand)
    counts = Dict('A' => 0, 'C' => 0, 'G' => 0, 'T' => 0)

    for nucleotide in strand
        haskey(counts, nucleotide) || throw(DomainError("Invalid nucleotide"))
        counts[nucleotide] += 1
    end

    counts
end
